### ISSUE TITLE
assertEqual is hard to read.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2893]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2893|
|Type|enhancement|
|Created|2007-11-13 02:47:31Z|
|Branch|https://github.com/twisted/twisted/tree/assertEqual-2893|

We should do what bzrlib does.

<details><summary>Searchable metadata</summary>

```
trac-id__2893 2893
type__enhancement enhancement
reporter__jml jml
priority__highest highest
milestone__ 
branch__branches_assertEqual_2893 branches/assertEqual-2893
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1194922051000000 1194922051000000
changetime__1194966404000000 1194966404000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

(In [21722]) Branching to 'assertEqual-2893'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @jml|
|-|-|

  * in twisted/trial/test/test_assertions.py:
    * The new test method ```TestAssertions.test_failUnlessEqualMessage``` has no docstring.
    * The modified method ```_testUnequalPair``` on the same class also has no docstring.
  * in twisted/trial/unittest.py:
    * The docstring of ```failUnlessEqual``` used to say _if msg is None, ..._.  But if ```None``` is passed for ```msg``` now, the output is a bit ugly:
```
twisted.trial.unittest.FailTest: Nonenot equal:
```
    maybe this case needs some test coverage?

The change is definitely an improvement.  Something to think about for a wider scope of improvement (ie, beyond the scope of this ticket) might be to leave this formatting up to the reporter and just construct the exception with the objects it needs (making them publically available on the exception).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml set owner to @exarkun|
|-|-|

Definitely agree regarding wider scope, but that's a research project.

Added docstring and a new test for msg=None.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @jml|
|-|-|

Great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml set status to `closed`|
|-|-|

(In [21730]) Provide a more readable error message when assertEqual fails.

 * Author: jml
 * Reviewer: exarkun
 * Fixes [#2893](https://github.com/twisted/twisted/issues/2893)

assertEqual's error message now always includes the objects being compared.
In addition, it displays these objects on separate lines and formats them
with pprint.pformat.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
